src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -2792,21 +2792,24 @@ protected override IPointerTypeSymbol CommonCreatePointerTypeSymbol(ITypeSymbol
             return CreatePointerTypeSymbol(elementType.EnsureCSharpSymbolOrNull<ITypeSymbol, TypeSymbol>("elementType"));
         }
 
-        protected override INamedTypeSymbol CommonCreateTupleTypeSymbol(ImmutableArray<ITypeSymbol> elementTypes, ImmutableArray<string> elementNames)
+        protected override INamedTypeSymbol CommonCreateTupleTypeSymbol(INamedTypeSymbol underlyingType, ImmutableArray<string> elementNames)
         {
-            if (elementTypes.IsDefault)
+            if ((object)underlyingType == null)
             {
-                throw new ArgumentNullException(nameof(elementTypes));
+                throw new ArgumentNullException(nameof(underlyingType));
             }
 
-            if (elementTypes.Length <= 1)
+            var x = underlyingType.EnsureCSharpSymbolOrNull<INamedTypeSymbol, NamedTypeSymbol>(nameof(underlyingType));
+
+            int cardinality;
+            if (!x.IsTupleCompatible(out cardinality))
             {
-                throw new ArgumentException(CodeAnalysisResources.TuplesNeedAtLeastTwoElements, nameof(elementNames));
+                throw new ArgumentException(CodeAnalysisResources.TupleUnderlyingTypeMustBeTupleCompatible, nameof(underlyingType));
             }
 
             if (!elementNames.IsDefault)
             {
-                if (elementNames.Length != elementTypes.Length)
+                if (elementNames.Length != cardinality)
                 {
                     throw new ArgumentException(CodeAnalysisResources.TupleNamesAllOrNone, nameof(elementNames));
```

Comment:
The message might be somewhat misleading when count of names is greater that cardinality. Consider adding a special message for that case.